File: src/main/java/com/blogly/blogly/infrastructure/persistence/user/UserJpaRepositoryAdapter.java
package com.blogly.blogly.infrastructure.persistence.user;

import com.blogly.blogly.domain.user.Email;
import com.blogly.blogly.domain.user.User;
import com.blogly.blogly.domain.user.UserId;
import com.blogly.blogly.domain.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@RequiredArgsConstructor
@Repository
public class UserJpaRepositoryAdapter implements UserRepository {

    private final UserJpaRepository repository;
    private final UserDomainMapper mapper;

    @Override
    public User findById(UserId id) {
        return repository.findById(id.getValue())
                .map(mapper::toDomain)
                .orElse(null);
    }

    @Override
    public User findByEmail(Email email) {
        return repository.findByEmail(email.getValue())
                .map(mapper::toDomain)
                .orElse(null);
    }

    @Override
    public boolean existsByEmail(Email email) {
        return repository.existsByEmail(email.getValue());
    }

    @Override
    public UserId save(User user) {
        UserEntity entity = mapper.toEntity(user);
        repository.save(entity);
        return user.getId();
    }
}

File: src/main/java/com/blogly/blogly/infrastructure/persistence/user/UserJpaRepositoryAdapter.kt
package com.blogly.blogly.infrastructure.persistence.user

import com.blogly.blogly.domain.shared.PageQuery
import com.blogly.blogly.domain.shared.PageResult
import com.blogly.blogly.domain.user.Email
import com.blogly.blogly.domain.user.User
import com.blogly.blogly.domain.user.UserId
import com.blogly.blogly.domain.user.UserRepository
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Repository

@Repository
class UserJpaRepositoryAdapter(
    private val repository: UserJpaRepository
) : UserRepository {

    override fun findById(id: UserId): User? =
        repository.findByIdOrNull(id.value)
            ?.let { UserDomainMapper.toDomain(it) }

    override fun findByEmail(email: Email): User? =
        repository.findByEmail(email.value)
            ?.let { UserDomainMapper.toDomain(it) }

    override fun findAll(pageQuery: PageQuery): PageResult<User> {
        val pageable = PageRequest.of(pageQuery.page, pageQuery.size, Sort.by(Sort.Direction.DESC, "id"))
        val users = repository.findAll(pageable)

        return PageResult(
            users.content.map(UserDomainMapper::toDomain),
            pageQuery.page,
            pageQuery.size,
            users.totalElements
        )
    }

    override fun existsByEmail(email: Email): Boolean =
        repository.existsByEmail(email.value)

    override fun save(user: User): UserId {
        val entity = UserDomainMapper.toEntity(user)
        repository.save(entity)
        return user.id
    }
}
